cli/cli-client/src/main/java/org/infinispan/cli/patching/PatchTool.java

```diff
@@ -0,0 +1,474 @@
+   }
+
+   private List<PatchInfo> getInstalledPatches(Path target) {
+      Path patchesFile = target.resolve(PATCHES_DIR).resolve(PATCHES_FILE);
+      try (InputStream is = Files.newInputStream(patchesFile, StandardOpenOption.READ)) {
+         JavaType patchInfoList = mapper.getTypeFactory().constructCollectionType(List.class, PatchInfo.class);
+         return mapper.readValue(is, patchInfoList);
+      } catch (NoSuchFileException e) {
+         return new ArrayList<>();
+      } catch (IOException e) {
+         throw MSG.patchCannotRead(patchesFile, e);
+      }
+   }
+
+   private void writeInstalledPatches(Path target, List<PatchInfo> patches) {
+      try (OutputStream os = Files.newOutputStream(Files.createDirectories(target.resolve(PATCHES_DIR)).resolve(PATCHES_FILE), StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+         mapper.writeValue(os, patches);
+      } catch (IOException e) {
+         throw MSG.patchCannotWritePatchesFile(e);
+      }
+   }
+
+   private void createSinglePatch(String qualifier, Path source, Path target, Version targetVersion, Map<Path, ServerFile> targetFiles, FileSystem zipfs) throws IOException {
+      Version sourceVersion = getVersion(source);
+
+      // Ensure that the brand name coincides
+      String sourceBrand = sourceVersion.brandName();
+      String targetBrand = targetVersion.brandName();
+      if (!sourceBrand.equals(targetBrand)) {
+         throw MSG.patchIncompatibleProduct(sourceBrand, targetBrand);
+      }
+
+      if (sourceVersion.version().equals(targetVersion.version())) {
+         throw MSG.patchServerAndTargetMustBeDifferent(sourceVersion.version());
+      }
+
+      PatchInfo patchInfo = new PatchInfo(sourceBrand, sourceVersion.version(), targetVersion.version(), qualifier);
+
+      // Build a list of files in the old version
+      Map<Path, ServerFile> v1Files = getServerFiles(source);
+
+      // Compare the two file lists, generating a list of upgrade instructions
+      List<PatchOperation> operations = patchInfo.getOperations();
+      v1Files.forEach((k1, v1File) -> {
+         if (!targetFiles.containsKey(k1)) {
+            operations.add(PatchOperation.remove(v1File.getVersionedPath(), v1File.getDigest(), v1File.getPermissions()));
+         } else {
+            ServerFile targetFile = targetFiles.get(k1);
+            if (!v1File.getFilename().equals(targetFile.getFilename())) { // Different filename means upgrade
+               operations.add(PatchOperation.upgrade(v1File.getVersionedPath(), v1File.getDigest(), v1File.getPermissions(), targetFile.getVersionedPath(), targetFile.getDigest(), targetFile.getPermissions()));
+               addFileToZip(zipfs, target, targetFile);
+            } else { // Check contents
+               try {
+                  if (!compareFileContents(source.resolve(v1File.getVersionedPath()), target.resolve(targetFile.getVersionedPath()))) {
```

Comment:
can't you just compare the `digest` instead of doing a binary check?

Reply:
Very stupid of me :)